cpp/src/gandiva/gdv_function_stubs.cc

```diff
@@ -122,6 +123,133 @@ int32_t gdv_fn_populate_varlen_vector(int64_t context_ptr, int8_t* data_ptr,
+                                        bool value_validity,                            \
+                                        int32_t *out_length) {                          \
+    if (!value_validity) {                                                              \
+      *out_length = 0;                                                                  \
+      return "";                                                                        \
+    }                                                                                   \
+                                                                                        \
+    return gandiva::HashUtils::HashUsingSha256(context, value,                          \
+    value_length, out_length);                                                          \
+  }
+
+// Expand inner macro for all numeric types.
+#define SHA_NUMERIC_BOOL_DATE_PARAMS(INNER) \
+  INNER(int8)                          \
+  INNER(int16)                         \
+  INNER(int32)                         \
+  INNER(int64)                         \
+  INNER(uint8)                         \
+  INNER(uint16)                        \
+  INNER(uint32)                        \
+  INNER(uint64)                        \
+  INNER(float32)                       \
+  INNER(float64)                       \
+  INNER(boolean)                       \
+  INNER(date64)                        \
+  INNER(date32)                        \
+  INNER(time32)                        \
+  INNER(timestamp)
+
+
+// Expand inner macro for all numeric types.
+#define SHA_VAR_LEN_PARAMS(INNER) \
+  INNER(utf8)                     \
+  INNER(binary)
+
+
+SHA_NUMERIC_BOOL_DATE_PARAMS(SHA256_HASH_FUNCTION)
+SHA_VAR_LEN_PARAMS(SHA256_HASH_FUNCTION_BUF)
+
+SHA_NUMERIC_BOOL_DATE_PARAMS(SHA128_HASH_FUNCTION)
+SHA_VAR_LEN_PARAMS(SHA128_HASH_FUNCTION_BUF)
+
+#undef SHA_NUMERIC_BOOL_DATE_PARAMS
+#undef SHA_VAR_LEN_PARAMS
+
+// Add functions for decimal128
+GANDIVA_EXPORT
+const char* gdv_fn_sha256_decimal128(int64_t context, int64_t x_high, uint64_t x_low,
+                                     int32_t x_precision, int32_t x_scale,
+                                     gdv_boolean x_isvalid, int32_t *out_length) {
+  if (!x_isvalid) {
+      *out_length = 0;
+      return "";
+  }
+
+  const auto &decimal_128 = gandiva::BasicDecimal128(x_high, x_low);
+  return gandiva::HashUtils::HashUsingSha256(context, decimal_128.ToBytes().data(),
+                                             16, out_length);
+}
+
+GANDIVA_EXPORT
+const char* gdv_fn_sha128_decimal128(int64_t context, int64_t x_high, uint64_t x_low,
+                                     int32_t x_precision, int32_t x_scale,
+                                     gdv_boolean x_isvalid, int32_t *out_length) {
+  if (!x_isvalid) {
+    *out_length = 0;
+    return "";
+  }
+
+  const auto &decimal_128 = gandiva::BasicDecimal128(x_high, x_low);
```

Comment:
This seems like a strange formulation?  Why note BasicDecimal128 decimal_128(x_high, x_low);

Reply:
I changed to `const BasicDecimal128 decimal_128(x_high, x_low)`